In [1]:
!pip install transformers
!pip install datasets
!pip install peft
!pip install accelerate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 320.7/320.7 kB 4.7 MB/s eta 0:00:00a 0:00:01


In [43]:
!pip install torch

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [29]:
!pip install evaluate

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 3.0 MB/s eta 0:00:00


In [5]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from datasets import load_dataset
from peft import get_peft_config, get_peft_model, LoraConfig, TaskType
from transformers import TrainingArguments, Trainer

In [7]:
from huggingface_hub import login

login("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")

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


LOADING DATASET AND TOKENIZATION

In [13]:
dataset = load_dataset("glue", "sst2")

README.md:   0%|          | 0.00/35.3k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/3.11M [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/72.8k [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/148k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/67349 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/872 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1821 [00:00<?, ? examples/s]

In [14]:
dataset = dataset["train"].shuffle(seed=1)

train_data = dataset.select(range(16000)) 
test_data = dataset.select(range(16000, 18000))  


print(f"Training data size: {len(train_data)}")
print(f"Test data size: {len(test_data)}")

Training data size: 16000
Test data size: 2000


In [26]:
tokenizer.pad_token = tokenizer.eos_token
def preprocess_function(examples):
    return tokenizer(examples["sentence"], truncation=True, padding="max_length", max_length=64)


test_dataset = test_data.map(preprocess_function, batched=True)
print("ok")
test_dataset = test_dataset.remove_columns(['sentence', 'idx'])  # Remove unnecessary columns
print("hello")
test_dataset.set_format(type='torch')
print("done")
print(test_dataset[0])

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

ok
hello
done
{'label': tensor(1), 'input_ids': tensor([128000,    385,   2396,   2225,  15612,    323,  34292,    220, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001]), 'attention_mask': tensor([1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])}


In [40]:
train_dataset = train_data.map(preprocess_function, batched=True)
print("ok")
train_dataset = train_dataset.remove_columns(['sentence', 'idx']) 
print("hello")
train_dataset.set_format(type='torch')
print("done")
print(train_dataset[0])

ok
hello
done
{'label': tensor(1), 'input_ids': tensor([128000,     64,  15473,   3265,     88,    323,  14134,  41367,   1833,
          5352,    311,    279,  47880,    597,  46295,  39542,   2559,   5352,
           364,     82,    602,    364,     76,    279,    832,    430,    602,
          1390,    662,    220, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001,
        128001]), 'attention_mask': tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])}


LOADING TOKENIZER and MODEL

In [24]:
model_name = "meta-llama/Llama-3.2-1B"  
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=2)

Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [25]:
for param in model.parameters():
    param.requires_grad = False

In [30]:
import evaluate

# Load metrics
accuracy_metric = evaluate.load("accuracy")
precision_metric = evaluate.load("precision")
recall_metric = evaluate.load("recall")
f1_metric = evaluate.load("f1")


In [31]:
def compute_metrics(p):
    logits, labels = p
    preds = logits.argmax(axis=-1) 

    accuracy = accuracy_metric.compute(predictions=preds, references=labels)
    precision = precision_metric.compute(predictions=preds, references=labels, average="binary")
    recall = recall_metric.compute(predictions=preds, references=labels, average="binary")
    f1 = f1_metric.compute(predictions=preds, references=labels, average="binary")

    return {
        'accuracy': accuracy['accuracy'],
        'precision': precision['precision'],
        'recall': recall['recall'],
        'f1': f1['f1']
    }


In [36]:
tokenizer.pad_token = tokenizer.eos_token
model.config.pad_token_id = tokenizer.pad_token_id

ZERO SHOT EVALUATION

In [37]:

training_args = TrainingArguments(
    output_dir='./results',          
    per_device_eval_batch_size=8,    
    no_cuda=True,                    
    do_train=False,                 
    do_eval=True,                   
)


trainer = Trainer(
    model=model,                     
    args=training_args,              
    eval_dataset=test_dataset,       
    compute_metrics=compute_metrics  
)


results = trainer.evaluate()
print(results)

wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Using wandb-core as the SDK backend. Please refer to https://wandb.me/wandb-core for more information.
wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:

  ········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


{'eval_loss': 2.172877550125122, 'eval_model_preparation_time': 0.0026, 'eval_accuracy': 0.441, 'eval_precision': 0.5, 'eval_recall': 0.0017889087656529517, 'eval_f1': 0.0035650623885918, 'eval_runtime': 1303.0834, 'eval_samples_per_second': 1.535, 'eval_steps_per_second': 0.192}


TRAINING THE MODEL

In [50]:
import torch 
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=2)  
model.classifier = torch.nn.Linear(model.config.hidden_size, 2) 

model.config.pad_token_id = tokenizer.pad_token_id
for param in model.model.parameters():
    param.requires_grad = False  

for param in model.classifier.parameters():
    param.requires_grad = True  
model

Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


LlamaForSequenceClassification(
  (model): LlamaModel(
    (embed_tokens): Embedding(128256, 2048)
    (layers): ModuleList(
      (0-15): 16 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (k_proj): Linear(in_features=2048, out_features=512, bias=False)
          (v_proj): Linear(in_features=2048, out_features=512, bias=False)
          (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=2048, out_features=8192, bias=False)
          (up_proj): Linear(in_features=2048, out_features=8192, bias=False)
          (down_proj): Linear(in_features=8192, out_features=2048, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((2048,), eps=1e-05)
        (post_attention_layernorm): LlamaRMSNorm((2048,), eps=1e-05)
      )
   

In [51]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)
model.to(device)

cuda


LlamaForSequenceClassification(
  (model): LlamaModel(
    (embed_tokens): Embedding(128256, 2048)
    (layers): ModuleList(
      (0-15): 16 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (k_proj): Linear(in_features=2048, out_features=512, bias=False)
          (v_proj): Linear(in_features=2048, out_features=512, bias=False)
          (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=2048, out_features=8192, bias=False)
          (up_proj): Linear(in_features=2048, out_features=8192, bias=False)
          (down_proj): Linear(in_features=8192, out_features=2048, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((2048,), eps=1e-05)
        (post_attention_layernorm): LlamaRMSNorm((2048,), eps=1e-05)
      )
   

In [56]:

training_args = TrainingArguments(
    output_dir='./results',  
    eval_strategy="steps",  
    logging_dir='./logs',  
    logging_steps=10,  
    per_device_train_batch_size=8,  
    per_device_eval_batch_size=8,
    num_train_epochs=3,  
    save_steps=100,  
    save_total_limit=2, 
    load_best_model_at_end=True, 
    do_train=True,
    do_eval=True,
    report_to="tensorboard",  
    fp16=True  
)

In [57]:
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    compute_metrics=compute_metrics
)

# Start training
trainer.train()

/opt/conda/lib/python3.10/site-packages/accelerate/accelerator.py:494: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.scaler = torch.cuda.amp.GradScaler(**kwargs)
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py:79: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.device(device), torch.cuda.stream(stream), autocast(enabled=autocast_enabled):
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py:79: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.device(device), torch.cuda.stream(stream), autocast(enabled=autocast_enabled):
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py:79: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.device(device), torch.cuda.stream(stream), autocast(enabled=autocast_enabled):
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0

TrainOutput(global_step=3000, training_loss=0.43978213731447857, metrics={'train_runtime': 35745.0723, 'train_samples_per_second': 1.343, 'train_steps_per_second': 0.084, 'total_flos': 1.7937180168192e+16, 'train_loss': 0.43978213731447857, 'epoch': 3.0})

CALCULATION OF METRICS FOR FINETUNED MODEL

In [72]:
finetuned_results=trainer.evaluate()

print(finetuned_results)

/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py:79: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.device(device), torch.cuda.stream(stream), autocast(enabled=autocast_enabled):
/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '


{'eval_loss': 0.3938862979412079, 'eval_accuracy': 0.829, 'eval_precision': 0.8533697632058288, 'eval_recall': 0.8381037567084079, 'eval_f1': 0.845667870036101, 'eval_runtime': 103.5928, 'eval_samples_per_second': 19.306, 'eval_steps_per_second': 1.207, 'epoch': 3.0}


In [60]:
trainer.save_model("./final_model")
tokenizer.save_pretrained("./final_model")


('./final_model/tokenizer_config.json',
 './final_model/special_tokens_map.json',
 './final_model/tokenizer.json')

In [63]:
!pip install huggingface_hub


/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [65]:
from huggingface_hub import login

login(token="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")

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /root/.cache/huggingface/token
Login successful


PUSHING TO HUGGINGFACE

In [66]:
from transformers import AutoModelForSequenceClassification, AutoTokenizer

model = AutoModelForSequenceClassification.from_pretrained("./final_model")
tokenizer = AutoTokenizer.from_pretrained("./final_model")


model.push_to_hub("Anura0505/llama_3.2_1B_SST_model")
tokenizer.push_to_hub("Anura0505/llama_3.2_1B_SST_tokenizer")


Some weights of the model checkpoint at ./final_model were not used when initializing LlamaForSequenceClassification: ['classifier.bias', 'classifier.weight']
- This IS expected if you are initializing LlamaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing LlamaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


model.safetensors:   0%|          | 0.00/4.94G [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

CommitInfo(commit_url='https://huggingface.co/Anura0505/llama_3.2_1B_SST_tokenizer/commit/3ea69b787abd31c3e9d01678aef603b8d21fd2ea', commit_message='Upload tokenizer', commit_description='', oid='3ea69b787abd31c3e9d01678aef603b8d21fd2ea', pr_url=None, repo_url=RepoUrl('https://huggingface.co/Anura0505/llama_3.2_1B_SST_tokenizer', endpoint='https://huggingface.co', repo_type='model', repo_id='Anura0505/llama_3.2_1B_SST_tokenizer'), pr_revision=None, pr_num=None)

COUNTING PARAMETERS IN PRETRAINED AND FINETUNED MODEL

In [67]:
def count_params(model):
    return sum(p.numel() for p in model.parameters())


print("total parameters in finetuned model:", count_params(model))

total parameters in finetuned model: 1235818496


In [68]:
print("Total parameters in pretrained model:", count_params(AutoModelForSequenceClassification.from_pretrained("meta-llama/Llama-3.2-1B", num_labels=2)))

Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Total parameters in pretrained model: 1235818496


In [71]:
print(model==AutoModelForSequenceClassification.from_pretrained("meta-llama/Llama-3.2-1B", num_labels=2))


Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


False


Q7)A) and Q7)c)

Here's the comparison table for the pretrained and finetuned models:

| Metric       | Pretrained Model  | Finetuned Model |
|--------------|-------------------|-----------------|
| Loss         | 2.1729            | 0.3939          |
| Accuracy     | 44.1%             | 82.9%           |
| Precision    | 50.0%             | 85.34%          |
| Recall       | 0.18%             | 83.81%          |
| F1 Score     | 0.36%             | 84.57%          |

Key Observations:
1. The finetuned model shows significant improvement across all metrics compared to the pretrained model.
2. Precision, recall, and F1 score show a massive leap, indicating the model performs better in capturing relevant instances after fine-tuning.

Q5) and Q7)B) 

Though the number of parameters for both the models are same, the classifier layer has different values of weights. The finetuned model obtains its weights through training on the SST2 dataset while the pretrained model has the weights randomly initialized. This is the reason why the finetuned model performa so much better than the pretrained model(zero-shot).